pkg/storaged/resize.jsx

```diff
@@ -0,0 +1,441 @@
+        fsys = client.blocks_fsys[cleartext.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(cleartext);
+        stratis_bdev = client.blocks_stratis_blockdev[cleartext.path];
+        if (crypto.MetadataSize !== undefined)
+            crypto_overhead = crypto.MetadataSize;
+        else
+            crypto_overhead = block.Size - cleartext.Size;
+    } else {
+        fsys = client.blocks_fsys[block.path];
+        vdo = client.legacy_vdo_overlay.find_by_backing_block(block);
+        stratis_bdev = client.blocks_stratis_blockdev[block.path];
+        crypto_overhead = 0;
+    }
+
+    function fsys_resize() {
+        if (fsys) {
+            // HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1934567
+            //
+            // block_fsys.MountedAt might be out of synch with reality
+            // here if resizing the crypto container accidentally
+            // triggered an unmount.  Thus, we check synchronously
+            // whether or not we should be doing a offline resize or
+            // not.
+            //
+            // Another option for us would be to just mount the
+            // filesystem back if that's what we expect, to undo the
+            // bug mentioned above. But let's be a bit more passive
+            // here and hope the bug gets fixed eventually.
+            return (is_mounted_synch(client.blocks[fsys.path])
+                    .then(is_mounted => {
+                        // When doing an offline resize, we need to first repair the filesystem.
+                        if (!is_mounted) {
+                            return (fsys.Repair({ })
+                                    .then(function () { return fsys.Resize(size - crypto_overhead, { }) }));
+                        } else {
+                            return fsys.Resize(size - crypto_overhead, { });
+                        }
+                    }));
+        } else if (vdo) {
+            if (size - crypto_overhead > vdo.physical_size)
+                return vdo.grow_physical();
+            else if (size - crypto_overhead < vdo.physical_size)
+                return Promise.reject(_("VDO backing devices can not be made smaller"));
+            else
+                return Promise.resolve();
+        } else if (stratis_bdev) {
+            const delta = size - crypto_overhead - Number(stratis_bdev.TotalPhysicalSize);
+            if (delta > 0) {
+                const pool = client.stratis_pools[stratis_bdev.Pool];
+                return pool.GrowPhysicalDevice(stratis_bdev.Uuid).then(std_reply);
+            } else if (delta < 0)
+                return Promise.reject(_("Stratis blockdevs can not be made smaller"));
+            else
+                return Promise.resolve();
+        } else if (size < orig_size) {
+            // This shouldn't happen.  But if it does, continuing is harmful, so we throw an error.
+            return Promise.reject(_("Unrecognized data can not be made smaller here."));
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19385-20230926-092314-7463d2dd-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)